compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java

```diff
@@ -1726,22 +1771,81 @@ private void addToBlockedList(BLangInvocation invocationExpr) {
         this.taintedStatus = TaintedStatus.UNTAINTED;
     }
 
+    private void combinedParameterTaintedStatus(List<TaintedStatus> combinedArgTaintedStatus,
+                                                List<Boolean> argTaintedStatus) {
+        if (argTaintedStatus != null) {
+            if (combinedArgTaintedStatus.isEmpty()) {
+                List<TaintedStatus> argTaintedStatusTempList = argTaintedStatus.stream().map(taintedStatus ->
+                        taintedStatus ? TaintedStatus.TAINTED : TaintedStatus.UNTAINTED).collect(Collectors.toList());
+                combinedArgTaintedStatus.addAll(argTaintedStatusTempList);
+            } else {
+                // Merge lists together while making sure "Tainted" status of "argTaintedStatus" list is not overwritten
+                // with "Untainted", whereas it is allowed to overwritten "Untainted" status with with "Tainted".
+                for (int paramIndex = 0; paramIndex < argTaintedStatus.size(); paramIndex++) {
+                    if (argTaintedStatus.get(paramIndex) == Boolean.TRUE) {
+                        combinedArgTaintedStatus.set(paramIndex, TaintedStatus.TAINTED);
+                    }
+                }
+            }
+        }
+    }
+
+    private void updateParameterTaintedStatuses() {
+        updateParameterTaintedStatuses(requiredParams, 0);
+        updateParameterTaintedStatuses(defaultableParams, requiredParams.size());
+        if (restParam != null) {
+            updateParameterTaintedStatuses(Collections.singletonList(restParam),
+                    requiredParams.size() + defaultableParams.size());
+        }
+    }
+
+    private void updateParameterTaintedStatuses(List<BLangVariable> paramList, int startIndex) {
+        if (parameterTaintedStatus.size() <= startIndex) {
+            paramList.forEach(param -> {
+                if (hasAnnotation(param.annAttachments, ANNOTATION_TAINTED)) {
+                    parameterTaintedStatus.add(TaintedStatus.TAINTED);
+                } else {
+                    parameterTaintedStatus.add(param.symbol.tainted ?
+                            TaintedStatus.TAINTED : TaintedStatus.UNTAINTED);
+                }
+            });
+        } else {
+            for (int paramIndex = 0; paramIndex < paramList.size(); paramIndex++) {
+                BLangVariable param = paramList.get(paramIndex);
+                if (param.symbol.tainted) {
+                    parameterTaintedStatus.set(startIndex + paramIndex, param.symbol.tainted ?
+                            TaintedStatus.TAINTED : TaintedStatus.UNTAINTED);
```

Comment:
I think we can simplify this line with `parameterTaintedStatus.set(startIndex + paramIndex, TaintedStatus.TAINTED);`